### PR TITLE
KAFKA-2825, KAFKA-2851: Controller failover tests added to ducktape replication tests and fix to temp dir

### PR DESCRIPTION
**Pull request by apovzner**
I closed an original pull request that contained previous comments by Geoff (which are already addressed here), because I got into bad rebase situation. So, I created a new branch and cherry-picked my commits + merged with Ben's changes to fix MiniKDC tests to run on Virtual Box. That change was conflicting with my changes, where I was copying MiniKDC files with new scp method, and temp file was created inside that method. To merge Ben's changes, I added two optional parameters to scp(): 'pattern' and 'subst' to optionally substitute string while spp'ing files, which is needed for krb5.conf file.

### CODE REVIEW

File: tests/kafkatest/services/security/security_config.py [hunk 1]
```diff
@@ -110,7 +111,7 @@ def __init__(self, security_protocol, interbroker_security_protocol=None, sasl_m
     def client_config(self, template_props=""):
         return SecurityConfig(self.security_protocol, sasl_mechanism=self.sasl_mechanism, template_props=template_props)
 
-    def setup_node(self, node):
+    def setup_node(self, node, miniKdc=None):
```
Comment (granders) [inline tests/kafkatest/services/security/security_config.py:114]: Does it make sense to set `self.kdc` in the constructor? And then having non-None self.kdc would be part of the logic in `has_sasl_kerberos`

File: tests/kafkatest/services/zookeeper.py [hunk 2]
```diff
@@ -83,3 +84,19 @@ def clean_node(self, node):
 
     def connect_setting(self):
         return ','.join([node.account.hostname + ':2181' for node in self.nodes])
+
+    def query(self, path):
```
Comment (granders) [inline tests/kafkatest/services/zookeeper.py:88]: This could use a docstring
e.g. `"""afsadfasd"""`

**Comment by granders** [comment]
@apovzner A few minor comments, and needs rebasing, but this looks good.

File: tests/kafkatest/services/security/security_config.py [hunk 3]
```diff
@@ -127,8 +128,9 @@ def setup_node(self, node):
             jaas_conf = self.render(jaas_conf_file,  node=node, is_ibm_jdk=is_ibm_jdk)
             node.account.create_file(SecurityConfig.JAAS_CONF_PATH, jaas_conf)
             if self.has_sasl_kerberos:
-                node.account.scp_to(MiniKdc.LOCAL_KEYTAB_FILE, SecurityConfig.KEYTAB_PATH)
-                node.account.scp_to(MiniKdc.LOCAL_KRB5CONF_FILE, SecurityConfig.KRB5CONF_PATH)
+                scp(miniKdc.nodes[0], MiniKdc.KEYTAB_FILE, node, SecurityConfig.KEYTAB_PATH)
```
Comment (granders) [inline tests/kafkatest/services/security/security_config.py]: Maybe just check that `minikdc` is not None here

**Comment by apovzner** [comment]
@granders I just pushed the changes to address your comments.

**Comment by ijuma** [comment]
Conflicts need resolving on this again unfortunately.

**Comment by ijuma** [comment]
@apovzner, it looks like this was merged via different PRs, would you mind closing this one please?

**Comment by apovzner** [comment]
Closed because the code in this PR was two KAFKA JIRAs so I separated it into two separate PRs and already merged.

**Comment by b-goyal** [comment]
I closed an original pull request that contained previous comments by Geoff (which are already addressed here), because I got into bad rebase situation. So, I created a new branch and cherry-picked my commits + merged with Ben's changes to fix MiniKDC tests to run on Virtual Box. That change was conflicting with my changes, where I was copying MiniKDC files with new scp method, and temp file was created inside that method. To merge Ben's changes, I added two optional parameters to scp(): 'pattern' and 'subst' to optionally substitute string while spp'ing files, which is needed for krb5.conf file.

**Pull request closed without merging**